### PR TITLE
Add parallel coset shift method (speeds up coset_fft), add fft correctness tests

### PR DESCRIPTION
**Pull request by ValarDragon**
## Description

This PR adds parallelization to computing the coset shift, which is important for coset_ffts. Furthermore, it adds tests to show that all radix 2 FFTs are actually correct. (This previously was not here)

The speed improvements of this change on a 48 core benchmark server:
```
"bls12_381 - radix2" - coset_fft_in_place/131072
                        time:   [33.022 ms 33.577 ms 34.128 ms]
                        change: [-31.177% -29.858% -28.472%] (p = 0.00 < 0.05)
                        Performance has improved.
"bls12_381 - radix2" - coset_fft_in_place/262144
                        time:   [60.894 ms 61.374 ms 61.877 ms]
                        change: [-20.505% -19.577% -18.747%] (p = 0.00 < 0.05)
                        Performance has improved.
"bls12_381 - radix2" - coset_fft_in_place/524288
                        time:   [110.27 ms 110.95 ms 111.67 ms]
                        change: [-19.993% -19.415% -18.737%] (p = 0.00 < 0.05)
                        Performance has improved.
"bls12_381 - radix2" - coset_fft_in_place/1048576
                        time:   [212.02 ms 212.96 ms 213.91 ms]
                        change: [-19.769% -19.319% -18.870%] (p = 0.00 < 0.05)
                        Performance has improved.
```

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (main)
- [x] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Wrote unit tests
- [x] Updated relevant documentation in the code
- [x] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

File: poly-benches/benches/fft.rs [hunk 1]
```diff
@@ -49,7 +49,7 @@ fn setup_bench(c: &mut Criterion, name: &str, bench_fn: fn(&mut Bencher, &usize)
 fn fft_common_setup<F: FftField, D: EvaluationDomain<F>>(degree: usize) -> (D, Vec<F>) {
     let mut rng = &mut rand::thread_rng();
     let domain = D::new(degree).unwrap();
-    let a = DensePolynomial::<F>::rand(degree, &mut rng)
+    let a = DensePolynomial::<F>::rand(degree - 1, &mut rng)
```
Comment (ValarDragon) [inline poly-benches/benches/fft.rs:52]: Driveby bug fix, this didn't affect the FFT time, but causes the meaning of the bench to have previously been incorrect.

File: poly/src/polynomial/univariate/dense.rs [hunk 2]
```diff
@@ -506,13 +506,16 @@ mod tests {
         for degree in 0..70 {
             let poly = DensePolynomial::<Fr>::rand(degree, &mut rng);
             let neg = -poly.clone();
-            assert!((poly + neg).is_zero());
+            let result = poly + neg;
+            assert!(result.is_zero());
+            assert_eq!(result.degree(), 0);
```
Comment (ValarDragon) [inline poly/src/polynomial/univariate/dense.rs:511]: drive by improvement to the additive identity tests. It would have caught the bug fixed by #119

**Review by Pratyush — APPROVED** [review]
